docs/api/main-config-indexers.md

```diff
@@ -0,0 +1,352 @@
+
+##### `importPath`
+
+(Required)
+
+Type: `string`
+
+The file to import from, e.g. the [CSF](./csf.md) file.
+
+It is likely that the [`fileName`](#filename) being indexed is not CSF, in which you will need to [transpile it to CSF](#transpiling-to-csf) so that Storybook can read it in the browser.
+
+##### `type`
+
+(Required)
+
+Type: `'story'`
+
+The type of entry.
+
+##### `metaId`
+
+Type: `string`
+
+Default: Auto-generated from [`title`](#title)
+
+Define the custom id for meta of the entry.
+
+If specified, the export default (meta) in the CSF file _must_ have a corresponding `id` property, to be correctly matched.
+
+##### `name`
+
+Type: `string`
+
+Default: Auto-generated from [`exportName`](#exportname)
+
+The name of the entry.
+
+##### `tags`
+
+Type: `string[]`
+
+Tags for filtering entries in Storybook and its tools.
+
+##### `title`
+
+Type: `string`
+
+Default: Auto-generated from default export of [`importPath`](#importpath)
+
+Determines the location of the entry in the sidebar.
+
+Most of the time, you should **not** specify a title, so that your indexer will use the default naming behavior. When specifying a title, you **must** use the [`makeTitle`](#maketitle) function provided in [`IndexerOptions`](#indexeroptions) to also use this behavior. For example, here's an indexer that merely appends a "Custom" prefix to the title derived from the file name:
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/main-config-indexers-title.js.mdx',
+    'common/main-config-indexers-title.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+##### `__id`
+
+Type: `string`
+
+Default: Auto-generated from [`title`](#title)/[`metaId`](#metaid) and [`exportName`](#exportname)
+
+Define the custom id for the story of the entry.
+
+If specified, the story in the CSF file **must** have a corresponding `__id` property, to be correctly matched.
+
+Only use this if you need to override the auto-generated id.
+
+## Transpiling to CSF
+
+The value of [`importPath`](#importpath) in an [`IndexInput`](#indexinput) must resolve to a [CSF](./csf.md) file. Most custom indexers, however, are only necessary because the input is _not_ CSF. Therefore, you will likely need to transpile the input to CSF, so that Storybook can read it in the browser and render your stories.
+
+Transpiling the custom source format to CSF is beyond the scope of this documentation. This transpilation is often done at the builder level ([Vite](../builders/vite.md) and/or [Webpack](../builders/webpack.md)), and we recommend using [unplugin](https://github.com/unjs/unplugin) to create plugins for multiple builders.
+
+The general architecture looks something like this:
+
+<!-- TODO: If diagram is useful, make a real one -->
+
+![Architecture diagram showing how custom indexers are used to generate stories from source files](./main-config-indexers-architecture.png)
+
+First, here's an example of a source file:
+
+```ts
```

Comment:
@kylegach or @JReinhold if this can be applied to both TS and JS variants and no specific types need to be added here we can probably leave inline and adjust the component comment to be something like `// Button.variants.js|ts`